packages/loader/container-loader/src/containerStateManager.ts

```diff
@@ -0,0 +1,209 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	ISequencedDocumentMessage,
+	ISnapshotTree,
+	IVersion,
+} from "@fluidframework/protocol-definitions";
+import { IGetPendingLocalStateProps, IRuntime } from "@fluidframework/container-definitions";
+import {
+	ITelemetryLoggerExt,
+	MonitoringContext,
+	PerformanceEvent,
+	UsageError,
+	createChildMonitoringContext,
+} from "@fluidframework/telemetry-utils";
+import { assert } from "@fluidframework/core-utils";
+import { IResolvedUrl, ISnapshot } from "@fluidframework/driver-definitions";
+import { isInstanceOfISnapshot } from "@fluidframework/driver-utils";
+import {
+	ContainerStorageAdapter,
+	ISerializableBlobContents,
+	getBlobContentsFromTree,
+} from "./containerStorageAdapter";
+import { IPendingContainerState } from "./container";
+
+export class containerStateManager {
+	private readonly savedOps: ISequencedDocumentMessage[] = [];
+	public snapshot:
+		| {
+				tree: ISnapshotTree;
+				blobs: ISerializableBlobContents;
+		  }
+		| undefined;
+	private readonly mc: MonitoringContext;
+	private readonly _offlineLoadEnabled: boolean;
+	private readonly storageAdapter: ContainerStorageAdapter;
+	private readonly isInteractiveClient: boolean;
+	private readonly pendingLocalState: IPendingContainerState | undefined;
+
+	constructor(
+		pendingLocalState: IPendingContainerState | undefined,
+		subLogger: ITelemetryLoggerExt,
+		private readonly getClientId: () => string | undefined,
+		private readonly getResolvedUrl: () => IResolvedUrl | undefined,
+		private readonly getRuntime: () => IRuntime | undefined,
+		storageAdapter,
+		isInteractiveClient,
+	) {
+		this.pendingLocalState = pendingLocalState;
+		this.mc = createChildMonitoringContext({
+			logger: subLogger,
+			namespace: "ContainerStateManager",
+		});
+		this.storageAdapter = storageAdapter;
+		this.isInteractiveClient = isInteractiveClient;
+		this._offlineLoadEnabled =
+			this.mc.config.getBoolean("Fluid.Container.enableOfflineLoad") ?? false;
+	}
+
+	public get offlineLoadEnabled(): boolean {
+		return this._offlineLoadEnabled;
+	}
+
+	public addSavedOp(message: ISequencedDocumentMessage) {
+		if (this.offlineLoadEnabled) {
+			this.savedOps.push(message);
+		}
+	}
+
+	public getSavedOps() {
+		return this.savedOps;
+	}
```

Comment:
this should be removed. saved op handling should be full encapsulated in this class
```suggestion

```